### ISSUE TITLE
Product schema

### ISSUE THREAD

**Issue by tonyhburns** [flashlightdb/flashlightdb#1]
TL;DR - We need an e-commerce product schema without most of the commerce bits.

Since we're using PostgreSQL and Elasticsearch, a semi-denormalized schema with judicious use of nested JSON structures is preferable here I think.

Here are the basic requirements as I see them:

- Ability to represent a hierarchy of main categories (Flashlight, Flashlight Part, Flashlight Part :arrow_right: LED, Accessories, Accessors :arrow_right: Pocket Clip, etc.)
  - Should products only belong to one main category?
- An additional taxonomy of tags that a product can have more than one of (e.g. Thrower, EDC, etc.)
- Products can have an arbitrary set of properties at the record level represented as a JSON object with nested arrays and objects as appropriate; CRUD views to edit products will be hard-coded in ActionView templates to represent different product types
- Primary key IDs of other objects are allowed inside of a product's properties object, and will be flattened out in Elasticsearch
- Need a way to represent both manufacturers and dealers, this can probably be a separate table or tables with an actual FK relationship with the products table
- Need to model the workflow state of a product (e.g. submitted for review, active in database, rejected, etc.)
- Are there any special aspects of the products we want in the database that would require special modelling in the schema?

Questions:

- How do we represent product variants? Should available colors be considered a full variant?
- How do we represent product editions?
